### PR TITLE
[Merged by Bors] - fix: provide `#min_imports` and other commands in `Mathlib.Init`

### PR DESCRIPTION
**Pull request by SnirBroshi**
Make these commands available across Mathlib:
- `#help`
- `#min_imports`
- `#min_imports in`
- `#find_home`
- `proof_wanted`

Add tests that verify commands work when importing basic files, to prevent regressions.

---
[Zulip](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/.23lint.20is.20broken/near/573303540)

I also think we should have `#check` (the tactic, not the Core command) and `#simp` and `#find_syntax`, but they import `Mathlib.Init` so it can't import them currently.

I'm not sure which of the above were available everywhere before the shake tool was run, let me know if I should add/remove commands.

~~Also there seems to be a bug in `#help` and the module system where it can't find the names of syntaxes, which affects the `#guard_msgs` in the test file, but it should be fixed independently of this PR.~~

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Init.lean [hunk 1]
```diff
@@ -24,7 +34,6 @@ public import Mathlib.Tactic.Linter.UnusedTactic
 public import Mathlib.Tactic.Linter.UnusedInstancesInType
 public import Mathlib.Tactic.Linter.Style
 public import Mathlib.Tactic.Linter.Whitespace
--- This import makes the `#min_imports` command available globally.
```
Comment (JovanGerb) [inline Mathlib/Init.lean:27]: This one gives the `#min_imports in` command, right? Could you write that in a comment?

File: Mathlib/Init.lean [hunk 2]
```diff
@@ -24,7 +34,6 @@ public import Mathlib.Tactic.Linter.UnusedTactic
...
 public import Mathlib.Tactic.MinImports
```
Comment (grunweg) [inline Mathlib/Init.lean]: This import is duplicate now --- please remove. (Does a linter cover this? There should be one; if not, that's another bug to fix.)

File: MathlibTest/Init.lean [hunk 3]
```diff
@@ -0,0 +1,44 @@
+module
+
+import Mathlib.Init
+
+/-! Checks that some utilities are available when importing `Mathlib.Init` alone -/
```
Comment (grunweg) [inline MathlibTest/Init.lean]:
```suggestion
/-! Checks that some utilities are available already when importing `Mathlib.Init`. -/
```
maybe (what do you think)?

**Review by grunweg — COMMENTED** [review]
Thanks for doing this! I also like Thomas' suggestion, please follow it as well.

I'm not as convinced about adding #loogle. Having small imports in Mathlib.Init is also reasonable, and #loogle seems less clearly useful to me. (If you disagree, that's also fine - then I'd like to ask more reviewers for their opinion.)

File: Mathlib/Init.lean (hunk 2 above)
Comment (grunweg) [reply Mathlib/Init.lean]: Never mind, it is different. Please move all these "imports for a specific commands" together, though.

File: Mathlib/Init.lean [hunk 4]
```diff
@@ -1,7 +1,17 @@
 module  -- shake: keep-all, shake: keep-downstream
 
+-- This import makes the `#help` command available globally.
+public import Batteries.Tactic.HelpCmd
+-- This import makes the `proof_wanted` command available globally.
+public import Batteries.Util.ProofWanted
+-- This import makes the `#find_home` command available globally.
+public import ImportGraph.Tools.FindHome
+-- This import makes the `#min_imports` command available globally.
+public import ImportGraph.Tools.MinImports
```
Comment (joneugster) [inline Mathlib/Init.lean]:
```suggestion
public import ImportGraph.Tools
```
does this work, too? That would be preferred IMO as it imports all `#xy`-commands defined in this package.

If it doesn't work, then I should probably look at the required fix necessary (in `ImportGraph`) to make it work...

File: Mathlib/Init.lean (hunk 2 above)
Comment (SnirBroshi) [reply Mathlib/Init.lean]: ~~It was duplicate before too, assuming you mean the chain `Mathlib.Init → Mathlib.Tactic.TacticAnalysis.Declarations → Mathlib.Tactic.ExtractGoal → Mathlib.Tactic.MinImports`.
Otherwise, you might be confusing it with `ImportGraph.Tools.MinImports`.~~
Edit: Sorry, didn't see your second comment before commenting this

File: Mathlib/Init.lean (hunk 4 above)
Comment (SnirBroshi) [reply Mathlib/Init.lean]: It does work but I tried to import as little as possible. If people like the other commands as well then sure

File: MathlibTest/Init.lean (hunk 3 above)
Comment (SnirBroshi) [reply MathlibTest/Init.lean]: Both sound a bit strange in their own way 😅

File: Mathlib/Init.lean (hunk 2 above)
Comment (grunweg) [reply Mathlib/Init.lean]: I had confused the imports... time to stop working and go to sleep :-)

**Review by joneugster — APPROVED** [review]
Thank you for the PR.

I was a bit worried about bloating `Init.lean`, but I think it does make sense if something like `#lint`, `#help` or `#min_imports` is available between `Init.lean` and `Tactic/Common.lean`, too.

maintainer merge

File: MathlibTest/BasicFiles/TacticCommon.lean [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+-- Lean.Compiler.LCNF.UnreachableBranches.functionSummariesExt extension: 2 new entries
+
+-- Lean.extraModUses extension: 1 new entries
+
+-- Lean.Parser.parserExtension extension: 3 new entries
+
+-- Lean.Elab.macroAttribute extension: 1 new entries
+-/
+#guard_msgs in
```
Comment (adomani) [inline MathlibTest/BasicFiles/TacticCommon.lean]: This guarded message here seems to expose too many implementation details, making this a maintenance burden.  Could you find some test that is independent of changes outside of mathlib?

File: MathlibTest/BasicFiles/TacticCommon.lean [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+whatsnew in
+notation "test_whatsnew" => 0
+
+/-- info: Used 3 heartbeats, which is less than the current maximum of 200000. -/
+#guard_msgs in
+#count_heartbeats in
```
Comment (adomani) [inline MathlibTest/BasicFiles/TacticCommon.lean:27]: Similar maintenance comment here: I think that there is an `approx` flag for `#count_heartbeats`.

File: MathlibTest/BasicFiles/TacticCommon.lean [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+/-! Checks that some utilities are available already when importing `Mathlib.Tactic.Common`. -/
+
+/-- info: Loogle Usage -/
+#guard_msgs (substring := true) in
+#loogle
+
+/-- Nat.add: ℕ → ℕ → ℕ -/
+#guard_msgs (substring := true) in
+#find Nat → Nat → Nat
```
Comment (adomani) [inline MathlibTest/BasicFiles/TacticCommon.lean:11]: Could you `#find` a more function that is maybe defined in this file, again for robustness?

File: MathlibTest/BasicFiles/Tactic.lean [hunk 8]
```diff
@@ -0,0 +1,14 @@
+module
+
+import Mathlib.Tactic
+
+/-! Checks that some utilities are available already when importing `Mathlib.Tactic`. -/
+
+/-- info: Found 5 uses among over 2000 syntax declarations -/
+#guard_msgs (substring := true) in
+#find_syntax "∘" approx
```
Comment (adomani) [inline MathlibTest/BasicFiles/Tactic.lean]: Another maintenance comment: could you test a syntax defined in mathlib, maybe even in this file?

File: MathlibTest/BasicFiles/Init.lean [hunk 9]
```diff
@@ -0,0 +1,54 @@
+module
+
+import Mathlib.Init
+
+/-! Checks that some utilities are available already when importing `Mathlib.Init`. -/
+
+/-- info: syntax "exact"... [Lean.Parser.Tactic.exact] -/
+#guard_msgs (substring := true) in
+#help tactic exact
```
Comment (adomani) [inline MathlibTest/BasicFiles/Init.lean]: Here also, could you use some syntax that is defined in mathlib?

**Review by adomani — COMMENTED** [review]
Overall, I am happy with the import changes.

However, I find that the tests can break too easily with changes *upstream* of mathlib.

Could you find tests that depend minimally (ideally not at all) on changes outside of mathlib?

**Review by Vierkantor — APPROVED** [review]
Thanks!

bors r+

**Review by adomani — COMMENTED** [review]
Thanks, this is indeed what I had in mind!  I also think that the tests could have used `(drop info)` or something similar, instead of matching on the empty string.

Anyway, it's merged now!  🙂
